src/Elmish.WPF/DynamicViewModel.fs

```diff
@@ -34,6 +34,75 @@ module internal IViewModel =
+    Bindings: IReadOnlyDictionary<string, VmBinding<'model, 'msg, obj>>
+    ValidationErrors: IReadOnlyDictionary<string, string list ref>
+    PropertyChanged: Event<PropertyChangedEventHandler, PropertyChangedEventArgs>
+    ErrorsChanged: DelegateEvent<EventHandler<DataErrorsChangedEventArgs>> }
+
+  interface INotifyPropertyChanged with
+    [<CLIEvent>]
+    member x.PropertyChanged = x.PropertyChanged.Publish
+
+  interface INotifyDataErrorInfo with
+    [<CLIEvent>]
+    member x.ErrorsChanged = x.ErrorsChanged.Publish
+    member x.HasErrors =
+      // WPF calls this too often, so don't log https://github.com/elmish/Elmish.WPF/issues/354
+      x.ValidationErrors
+      |> Seq.map (fun (Kvp(_, errors)) -> errors.Value)
+      |> Seq.filter (not << List.isEmpty)
+      |> (not << Seq.isEmpty)
+    member x.GetErrors name =
+      let name = name |> Option.ofObj |> Option.defaultValue "<null>" // entity-level errors are being requested when given null or ""  https://docs.microsoft.com/en-us/dotnet/api/system.componentmodel.inotifydataerrorinfo.geterrors#:~:text=null%20or%20Empty%2C%20to%20retrieve%20entity-level%20errors
+      x.LoggingArgs.log.LogTrace("[{BindingNameChain}] GetErrors {BindingName}", x.LoggingArgs.nameChain, name)
+      x.ValidationErrors
+      |> IReadOnlyDictionary.tryFind name
+      |> Option.map (fun errors -> errors.Value)
+      |> Option.defaultValue []
+      |> (fun x -> upcast x)
+
+module internal ViewModelHelper =
+
+  let create getSender args bindings validationErrors ={
+    GetSender = getSender
+    LoggingArgs = args.loggingArgs
+    Model = args.initialModel
+    ValidationErrors = validationErrors
+    Bindings = bindings
+    PropertyChanged = Event<PropertyChangedEventHandler, PropertyChangedEventArgs>()
+    ErrorsChanged = DelegateEvent<EventHandler<DataErrorsChangedEventArgs>>()
+  }
+
+  let updateModel newModel helper =
+    let {
+      log = log
+      nameChain = nameChain } = helper.LoggingArgs
+
+    let raisePropertyChanged name =
+      log.LogTrace("[{BindingNameChain}] PropertyChanged {BindingName}", nameChain, name)
+      helper.PropertyChanged.Trigger(helper.GetSender (), PropertyChangedEventArgs name)
+    let raiseCanExecuteChanged (cmd: Command) =
+      cmd.RaiseCanExecuteChanged ()
+    let raiseErrorsChanged name =
+      log.LogTrace("[{BindingNameChain}] ErrorsChanged {BindingName}", nameChain, name)
+      helper.ErrorsChanged.Trigger([| helper.GetSender (); box <| DataErrorsChangedEventArgs name |])
+
+    let eventsToRaise =
+      helper.Bindings
+      |> Seq.collect (fun (Kvp (name, binding)) -> Update(helper.LoggingArgs, name).Recursive(helper.Model, newModel, binding))
+      |> Seq.toList
+    eventsToRaise
+    |> List.iter (function
+      | ErrorsChanged name -> raiseErrorsChanged name
+      | PropertyChanged name -> raisePropertyChanged name
+      | CanExecuteChanged cmd -> cmd |> raiseCanExecuteChanged)
+
+    { helper with Model = newModel }
```

Comment:
Is `Model` the only field that changes?  If so, then I would prefer to emphasize this.  For example, maybe it would be best to have the model stored separately like it was before this PR.

Reply:
No, the `Bindings` and `ValidationErrors` fields will change as well.

Reply:
Added a PR to remove `Model` from the helper